9082 educate users on mat mul precision

yaoshiang opened the pull request:
Adds runnable tutorial to teach users about mat mul precision.

Also includes new manual build instructions for runnable tutorials.

yaoshiang commented:
PDF of doc is here 
[Controlling Floating Point Precision — PyTorch_XLA master documentation.pdf](https://github.com/user-attachments/files/20071422/Controlling.Floating.Point.Precision.PyTorch_XLA.master.documentation.pdf)

yaoshiang commented on docs/README.md:
```diff
@@ -28,4 +28,45 @@ git push origin gh-pages
+PyTorch. For now, include manual instructions and check-in the output ipynb and MD files. 
+
+Note that the existing .gitattributes in this directory will prevent diffs and merges
+on the ipynb, which is illegible json. 
```
thanks, fixed.

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# |
+# | Google TPUs are built with matrix multiplication optimized on silicon
+# | in a physical module called a Matrix Multiply Unit or MXU.
+# | The underlying technology is a systolic array. To maintain speed,
```
thanks, agree. Done.

yaoshiang commented on docs/README.md:
```diff
@@ -28,4 +28,45 @@ git push origin gh-pages
+py2nb source/tutorials/precision_tutorial.py
+jupyter nbconvert --to notebook --execute --inplace source/tutorials/precision_tutorial.ipynb 
+# Ignore SIGTERM
+# Manually verify precision_tutorial.ipynb that the final line shows a delta, see below.
```
Thanks, I updated the text. I wanted to keep the ordering of these instructions sequential so I did keep the location.

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | in a physical module called a Matrix Multiply Unit or MXU.
+# | The underlying technology is a systolic array. To maintain speed,
+# | researchers identified an inexpensive tradeoff: running matmuls in
+# | a floating point format called BF16, instead of FP32. BF16 is less
```
I have redrafted this entire section to start with the requirements (high range low precision) and present bfloat16 as the solution.

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | in a physical module called a Matrix Multiply Unit or MXU.
+# | The underlying technology is a systolic array. To maintain speed,
+# | researchers identified an inexpensive tradeoff: running matmuls in
+# | a floating point format called BF16, instead of FP32. BF16 is less
```
Totally agree, only using bfloat16 now. thanks!

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | BF16 does however have the same range of FP32, and important
+# | improvement over FP16.
+# |
+# | Nvidia V100 and later GPUs also include specialized matrix multiplication
```
thanks, done.

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | TF32, which is really better understood as "BF19": same
+# | range as BF16 and FP32, with 19 total bits.
+# |
+# | What this means is that for both Google TPUs and Nvidia TensorCores,
```
nice

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# |
+# | ![bits](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/tf32-bf16-fp16-fp32.png)
+
+# | ## Higher precision math on lower precision hardware
```
that doc doesn't share the math of how the 3 and 6 pass methods work, and we have had customer questions, so I think it's important we keep this description.

mikegre-google replied:
OK, sounds good. Let me take another pass at it.

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | resulting in $8.372e-1$. This is conceptually the "default" precision
+# | setting of PyTorch/XLA on TPU.
+# |
+# | The next approach is to break each number up into two parts,
```
mantissa and exponent are indeed how you break up an FP, but this is a different situation - it's more about breaking the mantissa up into different chunks to allow the 3 and 6 pass method.

yaoshiang commented on docs/runnable_tutorials:
thanks, removed unnecessary file.

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | This equals $(H \times H + H \times L + L \times H + L \times L)$.
+# | The fourth term is negligible, so it is ignored. The first three matrix
+# | multiplications comprise the three-pass approach and it roughly
+# | doubles the effective precision.
```
For this stylistic example in base-10, this would be almost 4 digits of base-10 precision.

yaoshiang replied:
but for bfloat16, it gets you around 15 digits of precision, not the full 23. The six pass I think gets you 22.5... almost 23, but, there's a corner case where the 24th bits are the equiv of (01 + 01 = 10) and "carries the one" to the left. 

but I have not been able to get a clear answer on this so I am reluctant to overcommit in this doc...

yaoshiang commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | two numbers with 4 digits of precision generates twice as many
+# | digits of precision in the result.
+# |
+# | The simplest approach is to round down the numbers to $9.1$ and $9.2$,
```
I think it does round to closest! I'll update.

yaoshiang commented on docs/README.md:
```diff
@@ -28,4 +28,45 @@ git push origin gh-pages
+
+Run every time: 
+
+py2nb source/tutorials/precision_tutorial.py
```
oh, absolutely - thanks for this one.

yaoshiang commented on docs/README.md:
```diff
@@ -28,4 +28,45 @@ git push origin gh-pages
+Note that the existing .gitattributes in this directory will prevent diffs and merges
+on the ipynb, which is illegible json. 
+
+TODO: Automate the docs build system for runnable tutorials. 
```
great call - https://github.com/pytorch/xla/issues/9136 and now in the README.

tengyifei commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+# | the differences between these three settings.
+# |
+# | Warning: Although this notebook demonstrates different precision settings,
+# | it is recommended to only set the precision once at the beginning of your
```
Not sure if it's already done; does it makes sense to throw a "hard error" when the user sets the precision twice?

yaoshiang replied:
It's a good question and I'm not sure. If I was pretty sure mat_mul_precision was flakey, I'd say, definitely yes. But I am also curious if the error is not really in the platform, but rather, in the testing harness. It seems weird to me that I can set matmul precision dynamically in scripts and interactive python interpreters... (see PR 9083)... but it fails the unit test. I really don't get it.

tengyifei replied:
Oh, I know why. Looks like our compilation caching is not sound.

I ran your test a bunch of times with different levels of matmul precision. Then I printed `xm.metrics_report()` and it only reported 1 compilation event. I think we need at least 3 compilation events if there are 3 different precision levels.

In contrast, JAX maintains an extensive context of ambient settings that will impact compilation results: https://github.com/jax-ml/jax/blob/35e2657be8308917c7fa407be5a0b53192134890/jax/_src/config.py#L230. Whenever _any_ of those things change, JAX will recompile.

I think the reasonable thing here is to:
- Print some warning when the precision level is changed, warning the user that existing cached graphs may nullify their precision level change.
- Someone should fix this in some follow-up.

As a corollary, probably we should advise users that they better change the precision once at the start of their script.

yaoshiang replied:
Thanks for this analysis! I suspected there was global state (because runtime didn't triple) but didn't know the tools you know to dig deeper. I couldn't bisect it between the python -m unittest runner and something in libtpu. 

Is there a deeper fix here for which I can file an issue? E.g. can we do what Jax does and force a recompile when this ambient setting is changed? Broadly speaking, being more stringent about what constitutes a cache hit? 

Either way, the note to only set this setting once is in this guide, and I'll add it to the doc string as well.

tengyifei replied:
> Is there a deeper fix here for which I can file an issue? E.g. can we do what Jax does and force a recompile when this ambient setting is changed? Broadly speaking, being more stringent about what constitutes a cache hit?

Absolutely. I'm only familiar with how JAX does it. But I could imagine a similar kind of `lazy_tensor_trace_context()` for PyTorch/XLA, and whatever dictionary we're currently inserting the cached compilation result into, we'll need to hash-combine the dict key with this trace context.

tengyifei replied:
It's probably a good idea to audit the JAX list and see if any other items apply to ptxla as well; not just the matmul precision

yaoshiang replied:
do we have the ability to force a cache flush? That would provide a workaround.

ysiraichi replied:
Maybe we could include the MatMul precision in the hash we use for retrieving already compiled graphs.

yaoshiang replied:
I've linked to this conversation in the new issue I filed on this topic, https://github.com/pytorch/xla/issues/9129

tengyifei commented on docs/source/tutorials/precision_tutorial.py:
```diff
@@ -0,0 +1,241 @@
+import struct
+
+
+def binary_fraction_to_fp32(bstr: str) -> float:
```
nit: maybe `assert bstr[:4] == "0b1."`

yaoshiang replied:
thanks, great idea. added a value error.

tengyifei reviewed (approved):
Thanks! I really enjoyed reading it.

yaoshiang commented:
Added logger.warning, added issue on examples I saw of logging.warning instead of logging.get_logger.warning, added unit test to assert the warning message, and consumed the warning message in all other unit tests to avoid polluting the unit test harness stdout.

yaoshiang commented:
I discussed w/ Mike Green off-line. Okay to merge now, he will open any issues for follow-up work.
